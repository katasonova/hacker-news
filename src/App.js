import React, { Component } from 'react';
import './App.css';

const DEFAULT_QUERY = 'redux';
const DEFAULT_HPP = '100';
const PATH_BASE = 'https://hn.algolia.com/api/v1';
const PATH_SEARCH = '/search';
const PARAM_SEARCH = 'query=';
const PARAM_PAGE = 'page=';
const PARAM_HPP = 'hitsPerPage=';

class App extends Component {

  constructor(props) {
    super(props);

    this.state = {
      results: null,
      searchKey: '',
      searchTerm: DEFAULT_QUERY,
      error: null
    }

    this.fetchSearchTopStories = this.fetchSearchTopStories.bind(this);
    this.onDismiss = this.onDismiss.bind(this);
    this.onSearchChange = this.onSearchChange.bind(this);
    this.setSearchTopStories = this.setSearchTopStories.bind(this);
    this.onSearchSubmit = this.onSearchSubmit.bind(this);
    this.needsToSearchTopStories = this.needsToSearchTopStories.bind(this);
  }

  fetchSearchTopStories(searchTerm, page = 0) {
    fetch(`${PATH_BASE}${PATH_SEARCH}?${PARAM_SEARCH}${searchTerm}&${PARAM_PAGE}${page}&${PARAM_HPP}${DEFAULT_HPP}`)
      .then(response => response.json())
      .then(result => this.setSearchTopStories(result))
      .catch(error => this.setState({ error }));
  }

  onDismiss(id) {
    const { searchKey, results } = this.state;
    const { hits, page } = results[searchKey];

    const isNotId = item => item.objectID !== id;
    const updatedHits = hits.filter(isNotId);

    this.setState({
      results: {
        ...results,
        [searchKey]: { hits: updatedHits, page }
      }
    })
  }

  onSearchChange(event) {
    this.setState({
      searchTerm: event.target.value
    })
  }

  needsToSearchTopStories(searchTerm) {
    return !this.state.results[searchTerm]
  }

  setSearchTopStories(result) {
    const { hits, page } = result;
    const { searchKey, results } = this.state;

    const oldHits = results && results[searchKey]
      ? results[searchKey].hits : [];

    const updatedHits = [...oldHits, ...hits];
    this.setState({
      results: {
        ...results,
        [searchKey]: { hits: updatedHits, page }
      }
    });
  }

  componentDidMount() {
    const { searchTerm } = this.state;
    this.setState({ searchKey: searchTerm });
    this.fetchSearchTopStories(searchTerm)
  }

  onSearchSubmit(event) {
    const { searchTerm } = this.state
    this.setState({ searchKey: searchTerm });

    if (this.needsToSearchTopStories(searchTerm)) {
      this.fetchSearchTopStories(searchTerm)
    }

    event.preventDefault();
  }

  //  вызывается каждый раз при изменении состояния компонента
  render() {
    const { searchKey, results, searchTerm, error } = this.state;
    const page = (results && results[searchKey] && results[searchKey].page) || 0;
    const list = (results && results[searchKey] && results[searchKey].hits) || [];

    // if (error) {
    //   return <p>Что-то произошло не так.</p>;
    // }

    return (
      <div className='page'>
        <div className='interactions'>
          <Search value={searchTerm} onChange={this.onSearchChange} onSubmit={this.onSearchSubmit}>Поиск</ Search>
        </div>
        {error ?
          <div className='interactions'>
            <p>Что-то произошло не так.</p>
          </div>
          : <Table list={list} onDismiss={this.onDismiss} />}

        <div className='interactions'>
          <Button onClick={() => {
            this.fetchSearchTopStories(searchKey, page + 1)
          }}>
            Больше историй
        </Button>
        </div>
      </div>
    );
  }
}

const Search = ({ value, onChange, onSubmit, children }) =>
  <form onSubmit={onSubmit}>
    <input type='text' value={value} onChange={onChange} />
    <button type='submit'>{children}</button>
  </form>

const Table = ({ list, onDismiss }) =>
  <div className='table'>
    {list.map((item =>
      <div key={item.objectID} className='table-row'>
        <span className='largeColumn'>
          <a href={item.url}>{item.title}</a>
        </span>
        <span className='midColumn'>{item.author}</span>
        <span className='smallColumn'>{item.num_comments}</span>
        <span className='smallColumn'>{item.points}</span>
        <Button onClick={() => onDismiss(item.objectID)}
          className='button-inline smallColumn'>
          Отбросить
          </Button>
      </div>
    )
    )}
  </div>


const Button = ({ onClick, className = '', children }) =>
  <button onClick={onClick} className={className} type='button'>
    {children}
  </button>

export default App;