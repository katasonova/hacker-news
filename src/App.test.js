import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import App, { Search, Button, Table } from './App';

describe('App', () => {

  it('Renders without errors', () => {
    const div = document.createElement('div');
    ReactDOM.render(<App />, div);
    ReactDOM.unmountComponentAtNode(div);
  })

  test('It is a corrrect snapshot', () => {
    const component = renderer.create(<App />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  })
});

describe('Search', () => {
  it('Renders without errors', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Search>Поиск</Search>, div);
    ReactDOM.unmountComponentAtNode(div);
  })

  test('It is a corrrect snapshot', () => {
    const component = renderer.create(<Search>Поиск</Search>);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  })
});

describe('Button', () => {
  it('Renders without errors', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Button>Больше историй</Button>, div);
    ReactDOM.unmountComponentAtNode(div);
  })

  test('It is a corrrect snapshot', () => {
    const component = renderer.create(<Button>Больше историй</Button>);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  })
});

describe('Table', () => {
  const props = {
    list: [
      { title: '1', author: '1', num_comments: 1, points: 2, objectID: 'y' },
      { title: '2', author: '2', num_comments: 1, points: 2, objectID: 'z' },
    ],
  };

  it('Renders without errors', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Table {...props} />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  test('It is a corrrect snapshot', () => {
    const component = renderer.create(<Table {...props} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  })

});