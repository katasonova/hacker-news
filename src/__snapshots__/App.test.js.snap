// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App It is a corrrect snapshot 1`] = `
<div
  className="page"
>
  <div
    className="interactions"
  >
    <form
      onSubmit={[Function]}
    >
      <input
        onChange={[Function]}
        type="text"
        value="redux"
      />
      <button
        type="submit"
      >
        Поиск
      </button>
    </form>
  </div>
  <div
    className="table"
  />
  <div
    className="interactions"
  >
    <div>
      Загрузка ...
    </div>
  </div>
</div>
`;

exports[`Button It is a corrrect snapshot 1`] = `
<button
  className=""
  type="button"
>
  Больше историй
</button>
`;

exports[`Search It is a corrrect snapshot 1`] = `
<form>
  <input
    type="text"
  />
  <button
    type="submit"
  >
    Поиск
  </button>
</form>
`;

exports[`Table It is a corrrect snapshot 1`] = `
<div
  className="table"
>
  <div
    className="table-row"
  >
    <span
      className="largeColumn"
    >
      <a>
        1
      </a>
    </span>
    <span
      className="midColumn"
    >
      1
    </span>
    <span
      className="smallColumn"
    >
      1
    </span>
    <span
      className="smallColumn"
    >
      2
    </span>
    <button
      className="button-inline smallColumn"
      onClick={[Function]}
      type="button"
    >
      Отбросить
    </button>
  </div>
  <div
    className="table-row"
  >
    <span
      className="largeColumn"
    >
      <a>
        2
      </a>
    </span>
    <span
      className="midColumn"
    >
      2
    </span>
    <span
      className="smallColumn"
    >
      1
    </span>
    <span
      className="smallColumn"
    >
      2
    </span>
    <button
      className="button-inline smallColumn"
      onClick={[Function]}
      type="button"
    >
      Отбросить
    </button>
  </div>
</div>
`;
